### ISSUE TITLE
init_args as dict

### ISSUE THREAD

**Issue by meatballs** [Axelrod-Python/Axelrod#706]
In the API, I think the strategies endpoint should indicate whether a strategy class takes init arguments and, if so, what those arguments should be.

In order to do so we'd need to re-work the way in which we currently handle init_args:
- A strategy class would have a dictionary whose keys give the names of the arguments (and, possibly, default values to be used if the arguments are omitted).
- On instantiation, the **init** method for the class would be passed keyword args (instead of args at present), which we'd need to compare against the class level dict.

**Comment by drvinceknight** [comment]
Would inspecting the init method of the class work? Using something like https://docs.python.org/2/library/inspect.html: which lets you "[...] extract and format the argument list for a function, or [...]"

**Comment by meatballs** [comment]
No - that's only going to work for instantiated objects. I want to get something from the class itself, so the API effectively says "here are all the strategy classes available to you and, for each one, everything you need to know in order to instantiate it"

**Comment by meatballs** [comment]
That would then allow a UI to present a list of strategies with enough information for a tournament to be defined and passed back to the API for execution.

**Comment by meatballs** [comment]
e.g. https://axelrod-api.herokuapp.com/strategies/random/

currently returns:
```
{
    "url": "https://axelrod-api.herokuapp.com/strategies/random/",
    "id": "random",
    "name": "Random",
    "description": "A player who randomly chooses between cooperating and defecting.",
    "classifier": {
        "makes_use_of": [],
        "manipulates_source": false,
        "stochastic": true,
        "manipulates_state": false,
        "long_run_time": false,
        "memory_depth": 0,
        "inspects_source": false
    }
```

which gives no indication that a probability should be passed. I'm suggesting that this should instead be:

```
{
    "url": "https://axelrod-api.herokuapp.com/strategies/random/",
    "id": "random",
    "name": "Random",
    "description": "A player who randomly chooses between cooperating and defecting.",
    "classifier": {
        "makes_use_of": [],
        "manipulates_source": false,
        "stochastic": true,
        "manipulates_state": false,
        "long_run_time": false,
        "memory_depth": 0,
        "inspects_source": false
    },
   "init_args": {
        "probability": 0
    }
```

**Comment by meatballs** [comment]
Possibly, the value for each dict entry should be a tuple of default value and a text description.

**Comment by meatballs** [comment]
or even a further dict containing:

* default
* description
* mandatory?

**Comment by drvinceknight** [comment]
I understand the use case.

My only current reservation with this idea is that (if I understand correctly) this would not be ideal for 'new to python contributors': they would no longer be writing a simple Python class but a simple Python class with a (given very simple) twist. We do already have an instance/player `init_args` so perhaps my reservation is not a good one.

FWIW I've just played with the inspect libary:

```
>>> import inspect
>>> class dumb_class:
...           def __init__(a, b, c):
...                  self.tup = (a, b, c)
>>> dumb_class.__init__.__code__.co_varnames
('a', 'b', 'c')
```

I tried it briefly with a strategy and it didn't work quite right (something perhaps to do with inheritance) but that's a 2 minute attempt so maybe there's something there?

**Comment by drvinceknight** [comment]
FYI (not directly related) it looks like the player `init_args` isn't setup quite right for the `Random` strategy anyway:

```
>>> s1 = axl.Random
>>> p1 = s1(p=.4)
>>> p1.init_args
()
```

If that's the case that could be pointing to a nasty bug right?

**Comment by meatballs** [comment]
Try:
```
>>> s1 = axl.Random
>>> p1 = s1(0.4)
>>> p1.init_args
(0.4,)
```

**Comment by meatballs** [comment]
>
My only current reservation with this idea is that (if I understand correctly) this would not be ideal for 'new to python contributors': they would no longer be writing a simple Python class but a simple Python class with a (given very simple) twist.

Actually, I think it would make things clearer and that normally leads to them being simpler.

For a class requiring init_args:

it would have an init_args dict at class level e.g.
```
init_args = {
    "probability": {
        "default": 0,
        "description": "The probability of the strategy defecting",
        "required": True
    }
}
````

and then its init method would either be:
```
def __init__(self, probability)
```
or:
```
def __init__(self, **init_args**)
```
(I prefer the latter)

We'd still use a decorator to handle it all and it would take care of checking that the dict passed to init matches the requirements of the class level init_args.

**Comment by drvinceknight** [comment]
> (I prefer the latter)

I think **I** prefer the latter too but from the pov of someone who has not written much Code/Python at all it is 'different' and would require an extra little step. Having said that, the library is evolving and it's probably not that big a hurdle for a newcomer.

It's also neat to have this further 'differentiation'/'clarification' between a class/instance and a strategy/player.

**Comment by meatballs** [comment]
I suspect that a class requiring init_args is already slightly more complex than a complete beginner would want to tackle.

**Comment by drvinceknight** [comment]
> I suspect that a class requiring init_args is already slightly more complex than a complete beginner would want to tackle.

I think you're missing an intermediate level of someone who has written a couple of text book classes but that's a moot point :) 👍

**Comment by meatballs** [comment]
Yeah, quite possibly. However, I think the danger of over-simplification rears its head here.

**Comment by drvinceknight** [comment]
As does over complication and abstraction. But just to clarify, my last couple of comments have been agreeing with you right? As I said I like the idea of differentiating the class/strategy and instance/player.

**Comment by meatballs** [comment]
We appear to be strenuously agreeing with one another. We should probably bring that to a halt!!

**Comment by drvinceknight** [comment]
:+1: :)

**Comment by meatballs** [comment]
I'll start having a look at this one - the API isn't going anywhere until this is sorted.

**Comment by meatballs** [comment]
Mostly resolved by #817 and superseded by #802
